chalk-solve/src/clauses.rs

```diff
@@ -167,6 +167,23 @@ fn program_clauses_that_could_match(
                 }
             }
 
+            // Check for impl and dyn Traits so that we generate
+            // `Implemented(impl Foo + Bar) :- Implemented(Foo), Implemented(Bar)`
+            match trait_ref.self_type_parameter() {
+                Some(Ty::Opaque(qwc)) | Some(Ty::Dyn(qwc)) => {
+                    clauses.push(ProgramClauseImplication {
+                        consequence: trait_ref.clone().cast(),
```

Comment:
This doesn't *quite* look right. For example, if this function were invoked with a `trait_ref` like `Implemented(dyn Foo: Bar)`, this would generate something like

```
implemented(dyn Foo: Bar) :-
    dyn Foo: Foo
```

but what we really want is 

```
Implemented(dyn Foo: Foo).
```

Moreover, something is a bit funny here -- you're just skipping over the binders in `qwc`, which isn't really right. What we want is more like:

```
let self_ty = trait_ref.self_type_parameter().unwrap(); // this should not really return an Option
let wc = qwc.instantiate(self_ty);
clauses.extend(wc.iter().cloned().casted());
```

what this will do is to take a `self_ty` like `dyn Foo` and create clauses like `Implemented(dyn Foo: Foo)`.

However, there are some complications.

For one, the `instantiated` method I referred to doesn't actually *exist* yet -- there is `subst::apply(&[self_ty], &qwc.value)`, which I think would be equivalent. The idea of `instantiate` would be to take a `Binders<T>` and create a `T`, substituting a value for the variables bound in `Binders`.